src/robin_dict.jl

```diff
@@ -0,0 +1,593 @@
+
+"""
+    RobinDict([itr])
+
+`RobinDict{K,V}()` constructs a hash table with keys of type `K` and values of type `V`.
+Keys are compared with [`isequal`](@ref) and hashed with [`hash`](@ref).
+Given a single iterable argument, constructs a [`RobinDict`](@ref) whose key-value pairs
+are taken from 2-tuples `(key,value)` generated by the argument.
+
+# Examples
+```jldoctest
+julia> RobinDict([("A", 1), ("B", 2)])
+RobinDict{String,Int64} with 2 entries:
+  "B" => 2
+  "A" => 1
+```
+
+Alternatively, a sequence of pair arguments may be passed.
+
+```jldoctest
+julia> RobinDict("A"=>1, "B"=>2)
+RobinDict{String,Int64} with 2 entries:
+  "B" => 2
+  "A" => 1
+```
+"""
+mutable struct RobinDict{K,V} <: AbstractDict{K,V}
+    #there is no need to maintain an table_size as an additional variable
+    # dibs::Array{Int8,1} # distance to initial bucket - critical for implementation
+    hashes::Vector{UInt32}
+    keys::Array{K,1}
+    vals::Array{V,1}
+    count::Int
+    totalcost::Int
+    maxprobe::Int   # length of longest probe
+    idxfloor::Int
+
+    function RobinDict{K, V}() where {K, V}
+        n = 16 # default size of an empty Dict in Julia
+        new(zeros(UInt32, n), Vector{K}(undef, n), Vector{V}(undef, n), 0, 0, 0, 0)
+    end
+
+    function RobinDict{K, V}(d::RobinDict{K, V}) where {K, V}
+        new(copy(d.hashes), copy(d.keys), copy(d.vals), d.count, d.totalcost, d.maxprobe, d.idxfloor)
+    end
+
+    function RobinDict{K, V}(keys, vals, hashes, count, totalcost, maxprobe, idxfloor) where {K, V}
+        new(hashes, keys, vals, count, totalcost, maxprobe, idxfloor)
+    end
+end
+
+function RobinDict{K,V}(kv) where V where K
+    h = RobinDict{K,V}()
+    for (k,v) in kv
+        h[k] = v
+    end
+    return h
+end
+RobinDict{K,V}(p::Pair) where {K,V} = setindex!(RobinDict{K,V}(), p.second, p.first)
+function RobinDict{K,V}(ps::Pair...) where V where K
+    h = RobinDict{K,V}()
+    sizehint!(h, length(ps))
+    for p in ps
+        h[p.first] = p.second
+    end
+    return h
+end
+
+RobinDict() = RobinDict{Any,Any}()
+RobinDict(kv::Tuple{}) = RobinDict()
+copy(d::RobinDict) = RobinDict(d)
+
+RobinDict(ps::Pair{K,V}...) where {K,V} = RobinDict{K,V}(ps)
+RobinDict(ps::Pair...)                  = RobinDict(ps)
+
+function RobinDict(d::AbstractDict{K, V}) where {K, V}
+    h = RobinDict{K, V}()
+    for (k, v) in d
+        h[k] = v
+    end
+    h
+end
+
+function RobinDict(kv)
+    try
+        d = dict_with_eltype((K, V) -> RobinDict{K, V}, kv, eltype(kv))
+        if isa(d, Dict)
+            return RobinDict(d)
+        else
+            return d
+        end
+    catch e
+    if !isiterable(typeof(kv)) || !all(x -> isa(x, Union{Tuple,Pair}), kv)
+            !all(x->isa(x,Union{Tuple,Pair}),kv)
+            throw(ArgumentError("RobinDict(kv): kv needs to be an iterator of tuples or pairs"))
+        else
+            rethrow(e)
+        end
+    end
+end
+
+hash_key(key) = ((hash(key)%UInt32) & 0xffffffff)|1
```

Comment:
`hash(key)` behaves according to architecture, it returns `UInt64` for 64 bit machines, `UInt32` otherwise. I am storing 32 bits hash, so I guess `(hash(key)%UInt32)|1` will be correct

Reply:
IIRC
The literal `1` depends on architecture also.
That function thus always returns that sized `Int`.

I recommend adding a test for this.

Reply:
From discussions with eulerkochy, I think our goal was to drop the high bit (ensuring the hash would never be exactly zero—our sentinel value for an empty cell, and easy to mask out later)
```suggestion
hash_key(key) = (hash(key) % UInt32) | 0x80000000
```
or we could just filter that value directly:
```suggestion
hash_key(key) = (h = hash(key); h == 0 ? UInt32(1) : h % UInt32)
```